observableeditablebuffer.go

```diff
@@ -0,0 +1,194 @@
+	for t := range e.observers {
+		tf(t)
+	}
+}
+
+// GetObserverSize will return the size of the observer map.
+func (e *ObservableEditableBuffer) GetObserverSize() int {
+	return len(e.observers)
+}
+
+// HasMultipleObservers returns true if their are multiple observers to the File.
+func (e *ObservableEditableBuffer) HasMultipleObservers() bool {
+	return len(e.observers) > 1
+}
+
+// MakeObservableEditableBuffer is a constructor wrapper for NewFile() to abstract File from the main program.
+func MakeObservableEditableBuffer(filename string, b RuneArray) *ObservableEditableBuffer {
+	f := NewFile(filename)
+	f.b = b
+	oeb := &ObservableEditableBuffer{
+		currobserver: nil,
+		observers:    nil,
+		f:            f,
+	}
+	oeb.f.oeb = oeb
+	return oeb
+}
+
+// MakeObservableEditableBufferTag is a constructor wrapper for NewTagFile() to abstract File from the main program.
+func MakeObservableEditableBufferTag(b RuneArray) *ObservableEditableBuffer {
+	f := NewTagFile()
+	f.b = b
+	oeb := &ObservableEditableBuffer{
+		currobserver: nil,
+		observers:    nil,
+		f:            f,
+	}
+	oeb.f.oeb = oeb
+	return oeb
+}
+
+// Clean is a forwarding function for file.Clean.
+func (e *ObservableEditableBuffer) Clean() {
+	e.f.Clean()
+}
+
+// FileSize is a forwarding function for file.Size.
+func (e *ObservableEditableBuffer) FileSize() int {
+	return e.f.Size()
+}
+
+// Mark is a forwarding function for file.Mark.
+func (e *ObservableEditableBuffer) Mark(seq int) {
+	e.f.Mark(seq)
+}
+
+// Reset is a forwarding function for file.Reset.
+func (e *ObservableEditableBuffer) Reset() {
+	e.f.Reset()
+}
+
+// HasUncommitedChanges is a forwarding function for file.HasUncommitedChanges.
+func (e *ObservableEditableBuffer) HasUncommitedChanges() bool {
+	return e.f.HasUncommitedChanges()
+}
+
+// IsDir is a forwarding function for file.IsDir.
+func (e *ObservableEditableBuffer) IsDir() bool {
+	return e.f.IsDir()
+}
+
+// SetDir is a forwarding function for file.SetDir.
+func (e *ObservableEditableBuffer) SetDir(flag bool) {
+	e.f.SetDir(flag)
+}
+
+// Nr is a forwarding function for file.Nr.
+func (e *ObservableEditableBuffer) Nr() int {
+	return e.f.Nr()
+}
+
+// ReadC is a forwarding function for file.ReadC.
+func (e *ObservableEditableBuffer) ReadC(q int) rune {
+	return e.f.ReadC(q)
+}
+
+// SaveableAndDirty is a forwarding function for file.SaveableAndDirty.
+func (e *ObservableEditableBuffer) SaveableAndDirty() bool {
+	return e.f.SaveableAndDirty()
+}
+
+// Load is a forwarding function for file.Load()
+func (e *ObservableEditableBuffer) Load(q0 int, fd io.Reader, sethash bool) (n int, hasNulls bool, err error) {
+	return e.f.Load(q0, fd, sethash)
+}
+
+// Dirty is a forwarding function for file.Dirty.
+func (e *ObservableEditableBuffer) Dirty() bool {
+	return e.f.Dirty()
+}
+
+// InsertAt is a forwarding function for file.InsertAt.
+func (e *ObservableEditableBuffer) InsertAt(p0 int, s []rune) {
+	e.AllObservers(func(i interface{}) {
```

Comment:
The order needs to be the same as the original `File.InsertAt`. You've altered the semantics. The change that you're telling the observers about hasn't happened yet.

Also: just loop over the observer map.

Reply:
My bad on the first part, and do you mean that I should replace the call to e.AllObservers here with a for loop?

Reply:
Yes. I meant that you could just do the normal `for` loop and not use `AllObservers`. `AllObservers` was me trying to be too cute.